### PR TITLE
Fix custom ads shadow draw on linux

### PR DESCRIPTION
**Pull request by aseren**
Chromium doesn't always support transparent window background on X11.
This can cause artifacts on shadows around ads notification popup. To fix
this shadows are drawn by Widget.

Before fix:
before
After fix:
after

Screenshot tests will be added in a separate PR.

Resolves https://github.com/brave/brave-browser/issues/15886

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave

  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: browser/ui/brave_ads/ad_notification_popup.cc [hunk 1]
```diff
@@ -406,7 +404,14 @@ void AdNotificationPopup::RecomputeAlignment() {
...
 
 const gfx::ShadowDetails& AdNotificationPopup::GetShadowDetails() const {
-  return gfx::ShadowDetails::Get(kShadowElevation, kCornerRadius);
+  // Do not draw shadow on Linux because shadows are already drawn by Widget.
+  const int shadow_elevation =
+#if defined(OS_LINUX)
+      0;
+#else
+      kShadowElevation;
+#endif
```
Comment (tmancey) [inline browser/ui/brave_ads/ad_notification_popup.cc:413]: `// defined(OS_LINUX)`

File: browser/ui/brave_ads/ad_notification_popup.cc [hunk 2]
```diff
@@ -51,21 +51,19 @@ namespace {
...
 
 constexpr SkColor kLightModeBorderColor = SkColorSetRGB(0xd5, 0xdb, 0xe2);
 constexpr SkColor kDarkModeBorderColor = SkColorSetRGB(0x3f, 0x41, 0x45);
-const int kBorderThickness = 1;
+constexpr int kBorderThickness = 1;
 
-#if defined(OS_WIN)
-const int kCornerRadius = 0;
+#if defined(OS_WIN) || defined(OS_LINUX)
+constexpr int kCornerRadius = 0;
```
Comment (tmancey) [inline browser/ui/brave_ads/ad_notification_popup.cc:64]: Does Linux not support a corner radius with shadows?

File: browser/ui/brave_ads/ad_notification_popup.cc [hunk 3]
```diff
@@ -422,7 +427,15 @@ void AdNotificationPopup::CreateWidgetView() {
   params.type = views::Widget::InitParams::TYPE_WINDOW_FRAMELESS;
   params.z_order = ui::ZOrderLevel::kFloatingWindow;
   params.opacity = views::Widget::InitParams::WindowOpacity::kTranslucent;
+#if defined(OS_LINUX)
+  // Some linux configurations do not support transparent content drawing. This
+  // can cause artifacts on shadows around ads notification popup.
+  // To fix this shadows are drawn by Widget.
+  params.shadow_type = views::Widget::InitParams::ShadowType::kDrop;
```
Comment (tmancey) [inline browser/ui/brave_ads/ad_notification_popup.cc:434]: I had rendering issues on Windows and macOS when the shadow was drawn by the Widget, any ideas if your change will work on macOS and Windows?

File: browser/ui/brave_ads/ad_notification_popup.cc (hunk 2 above)
Comment (aseren) [reply browser/ui/brave_ads/ad_notification_popup.cc:64]: No, setting the radius for widget has no effect.
If set the radius for background, it leads to folllowing artefacts:

Screenshot from 2021-07-02 20-31-47

File: browser/ui/brave_ads/ad_notification_popup.cc (hunk 3 above)
Comment (aseren) [reply browser/ui/brave_ads/ad_notification_popup.cc:434]: What kind of issues were drawn? I will check it on Windows.

File: browser/ui/brave_ads/ad_notification_popup.cc (hunk 1 above)
Comment (aseren) [reply browser/ui/brave_ads/ad_notification_popup.cc:413]: thanks!

File: browser/ui/brave_ads/ad_notification_popup.cc (hunk 3 above)
Comment (tmancey) [reply browser/ui/brave_ads/ad_notification_popup.cc:434]: The shadow was not always rendered until dragging the widget to another screen location

**Review by tmancey — APPROVED** [review]
LGTM++

File: browser/ui/brave_ads/window_util.cc [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+#include "ui/ozone/public/platform_gl_egl_utility.h"
+#endif
+
+#if defined(USE_X11)
+#include "ui/base/x/x11_util.h"
+#endif
+
+namespace brave_ads {
+
+bool IsTransparentWindowBackgroundSupported() {
```
Comment (aseren) [inline browser/ui/brave_ads/window_util.cc:20]: check code is written basing on this changes:
https://chromium-review.googlesource.com/c/chromium/src/+/2323293

File: browser/ui/brave_ads/ad_notification_popup.cc (hunk 3 above)
Comment (aseren) [reply browser/ui/brave_ads/ad_notification_popup.cc:431]: I didn't noticed such rendering issues on Windows. 
Shadow is drawn but InitParams::shadow_elevation and InitParams::corner_radius parameters are ignored.

Comment (tmancey) [reply browser/ui/brave_ads/ad_notification_popup.cc:431]: Once ready for review I will retest on Windows 10 and macOS. Thanks

**Comment by tmancey** [comment]
@aseren I have rebased and testing now

**Pull request merged**